### PR TITLE
Speed up image comparison

### PR DESCRIPTION
**Pull request by akx**
Closes #7260 (supersedes it).

I added a benchmark, and stacked the tests from the original (adjusted to pass lints, etc.) to be before the reimplementation in the PR, then added some more corner case tests.

I verified, with interactive rebase, that the tests pass just before the new handler is added, and then separately afterwards.

Local benchmark score improvement (unsurprising, and this isn't measuring peak memory usage improvements!):

```
------------------- benchmark 'compare': 15 tests, 2 sources -------------------
Name (time in ns)                          0001_main Min  0002_new Min      ΔMin
--------------------------------------------------------------------------------
test_equality[1237x811-L-inverted]          122,875.0916      470.7952    -99.6%
test_equality[1237x811-L-equal]             125,959.0499   24,207.9841    -80.8%
test_equality[1237x811-L-one-pixel]         129,332.9988   12,415.9269    -90.4%
test_equality[1237x811-I;16-one-pixel]      343,374.9080   27,542.0025    -92.0%
test_equality[1237x811-LA-inverted]         343,417.0503      479.1538    -99.9%
test_equality[1237x811-I;16-equal]          380,958.0812   55,041.9791    -85.6%
test_equality[1237x811-I;16-inverted]       387,958.9494      479.0956    -99.9%
test_equality[1237x811-LA-one-pixel]        451,375.0318  189,000.0422    -58.1%
test_equality[1237x811-LA-equal]            463,792.0065  378,541.9976    -18.4%
test_equality[1237x811-RGBA-inverted]       727,667.0076      483.2982    -99.9%
test_equality[1237x811-RGB-inverted]        754,792.0104      479.2004    -99.9%
test_equality[1237x811-RGB-equal]           813,374.9943  378,374.9416    -53.5%
test_equality[1237x811-RGB-one-pixel]       823,125.0104  189,290.9640    -77.0%
test_equality[1237x811-RGBA-equal]          874,208.0536  105,833.0527    -87.9%
test_equality[1237x811-RGBA-one-pixel]      899,208.0111   52,874.9079    -94.1%
--------------------------------------------------------------------------------
```

EDIT: added one more commit for autovectorisation to make RGB (likely very common!) and LA even faster (at least on my hardware):

```

--------------------- benchmark 'compare': 15 tests, 2 sources ---------------------
Name (time in ns)                          0001_main Min  0005_equalve Min      ΔMin
------------------------------------------------------------------------------------
test_equality[1237x811-L-inverted]          122,875.0916          479.0956    -99.6%
test_equality[1237x811-L-equal]             125,959.0499       24,333.0142    -80.7%
test_equality[1237x811-L-one-pixel]         129,332.9988       12,416.9746    -90.4%
test_equality[1237x811-I;16-one-pixel]      343,374.9080       27,625.0066    -92.0%
test_equality[1237x811-LA-inverted]         343,417.0503          479.2004    -99.9%
test_equality[1237x811-I;16-equal]          380,958.0812       55,207.9873    -85.5%
test_equality[1237x811-I;16-inverted]       387,958.9494          479.2004    -99.9%
test_equality[1237x811-LA-one-pixel]        451,375.0318       43,249.9219    -90.4%
test_equality[1237x811-LA-equal]            463,792.0065       86,874.9339    -81.3%
test_equality[1237x811-RGBA-inverted]       727,667.0076          485.3995    -99.9%
test_equality[1237x811-RGB-inverted]        754,792.0104          483.2982    -99.9%
test_equality[1237x811-RGB-equal]           813,374.9943       86,790.9985    -89.3%
test_equality[1237x811-RGB-one-pixel]       823,125.0104       43,459.0038    -94.7%
test_equality[1237x811-RGBA-equal]          874,208.0536      105,957.9663    -87.9%
test_equality[1237x811-RGBA-one-pixel]      899,208.0111       52,916.9338    -94.1%
------------------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by akx** [comment]
CodSpeed isn't showing a report here because all the benchmarks are new, but [you can see them in the report here](https://app.codspeed.io/python-pillow/Pillow/branches/akx%3Aimage-equals-new?q=equality).

The varied range of speeds suggests that these optimizations do take, _but_ I'm not sure if the last commit, the autovectorisation, works as well as it does on clang/ARM.

**Comment by hugovk** [comment]
Want to split the benchmarks to their own PR?

**Comment by akx** [comment]
@hugovk Done: #10033

File: src/PIL/Image.py [hunk 1]
```diff
@@ -754,13 +754,21 @@ def __eq__(self, other: object) -> bool:
...
+        # Early-out before realizing palettes.
+        # ImagingCore does check for mode and size.
+        if not (
             self.mode == other.mode
             and self.size == other.size
             and self.info == other.info
-            and self.getpalette() == other.getpalette()
-            and self.tobytes() == other.tobytes()
-        )
+        ):
```
Comment (radarhere) [inline src/PIL/Image.py]:
```suggestion
        if (
            self.mode != other.mode
            or self.size != other.size
            or self.info != other.info
        ):
```

File: src/PIL/Image.py [hunk 2]
```diff
@@ -754,13 +754,21 @@ def __eq__(self, other: object) -> bool:
...
+            return False
+        # Realize dirty palettes, if any.
+        # ImagingCore's compare function checks for palette equality.
```
Comment (radarhere) [inline src/PIL/Image.py]:
```suggestion
```
`load()` performs a far more important function in this context - it makes sure there is a core image present.

I'm not convinced we need these comments.

Comment (akx) [reply src/PIL/Image.py]: That too... 😄 Yeap, palette realization was just top of mind thanks to #9834.

File: src/PIL/Image.py [hunk 3]
```diff
@@ -754,13 +754,19 @@ def __eq__(self, other: object) -> bool:
...
-        return (
-            self.mode == other.mode
-            and self.size == other.size
-            and self.info == other.info
-            and self.getpalette() == other.getpalette()
-            and self.tobytes() == other.tobytes()
-        )
+        if self is other:
+            return True
+        # Early-out before realizing palettes.
```
Comment (radarhere) [inline src/PIL/Image.py]:
```suggestion
        # Early-out before loading.
```

**Comment by akx** [comment]
And the [memory usage comparison](https://gist.github.com/akx/4324a9843e9e2fe0170a35170a10f39d) that the benchmark doesn't capture shows that comparisons now require no extra allocations:

On a 2053x2039 image:

| Mode | main `==` (MiB) | new `==` (MiB) | main peak (MiB) | new peak (MiB) |
| --- | ---: | ---: | ---: | ---: |
| `1` | 3.7 | 0.0 | 11.8 | 8.1 |
| `CMYK` | 57.7 | 0.0 | 89.8 | 32.0 |
| `F` | 56.8 | 0.0 | 88.8 | 32.0 |
| `HSV` | 42.0 | 0.0 | 74.0 | 32.0 |
| `I` | 56.8 | 0.0 | 88.8 | 32.0 |
| `I;16` | 28.7 | 0.0 | 44.8 | 16.0 |
| `I;16B` | 39.9 | 0.0 | 55.9 | 16.0 |
| `I;16L` | 29.5 | 0.0 | 45.5 | 16.0 |
| `I;16N` | 28.7 | 0.0 | 44.8 | 16.0 |
| `L` | 15.2 | 0.0 | 23.3 | 8.1 |
| `LA` | 28.7 | 0.0 | 60.7 | 32.0 |
| `LAB` | 53.8 | 0.0 | 85.9 | 32.1 |
| `La` | 28.7 | 0.0 | 60.7 | 32.0 |
| `P` | 16.9 | 0.0 | 25.2 | 8.2 |
| `PA` | 28.8 | 0.0 | 61.0 | 32.1 |
| `RGB` | 42.0 | 0.0 | 74.0 | 32.0 |
| `RGBA` | 56.8 | 0.0 | 88.8 | 32.0 |
| `RGBX` | 56.8 | 0.0 | 88.8 | 32.0 |
| `RGBa` | 56.8 | 0.0 | 88.8 | 32.0 |
| `YCbCr` | 42.0 | 0.0 | 74.0 | 32.0 |

File: docs/releasenotes/13.0.0.rst [hunk 4]
```diff
@@ -193,3 +193,8 @@ Fixed histograms for masked LA, La and PA images
 When :py:meth:`~PIL.Image.Image.histogram` was given a mask for a two-band image (LA,
 La or PA), the second band was incorrectly derived from the first band of the input
 image, rather than the alpha band.
+
+Faster image comparison
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Comparing two images with ``==`` no longer copies both images out to ``bytes`` first.
```
Comment (radarhere) [inline docs/releasenotes/13.0.0.rst:200]: This might just be me, but I find this kind of vague. It doesn't copy the images to ``bytes`` first... but it still does eventually? What else does it do first?

Comment (akx) [reply docs/releasenotes/13.0.0.rst:200]: WDYT of 
```suggestion
Comparing two images with ``==`` now compares the in-memory representation of
the images, instead of copying their contents to new byte buffers to compare.
```

Is that too technical maybe?

```suggestion
Comparing two images with ``==`` was made faster by reducing data copying.
```

would at least be succinct?
